File: src/main/java/org/example/eshop/controller/ProductController.java
package org.example.eshop.controller;

import org.example.eshop.dto.ProductDetailDto;
import org.example.eshop.dto.ProductDto;
import org.example.eshop.service.ProductService;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import java.math.BigDecimal;
import java.util.List;

@RestController
@RequestMapping("/api/products")
public class ProductController {

    private final ProductService productService;

    public ProductController(ProductService productService) {
        this.productService = productService;
    }

    @GetMapping
    public ResponseEntity<List<ProductDto>> getProducts(
            @RequestParam(required = false) String type,
            @RequestParam(required = false) String region,
            @RequestParam(required = false) Integer harvestYear,
            @RequestParam(required = false) BigDecimal minPrice,
            @RequestParam(required = false) BigDecimal maxPrice,
            @RequestParam(required = false) Boolean inStock
    ) {
        try {
            List<ProductDto> products = productService.findProducts(
                    type,
                    region,
                    harvestYear,
                    minPrice,
                    maxPrice,
                    inStock
            );
            return ResponseEntity.ok(products);
        } catch (IllegalArgumentException e) {
            return ResponseEntity.badRequest().build();
        }
    }

    @GetMapping("/{slug}")
    public ResponseEntity<ProductDetailDto> getProductBySlug(@PathVariable String slug) {
        ProductDetailDto product = productService.findProductBySlug(slug);
        if (product != null) {
            return ResponseEntity.ok(product);
        }
        return ResponseEntity.notFound().build();
    }

    @GetMapping("/variant-info")
    public ResponseEntity<String> getVariantInfo(@RequestParam Long variantId) {
        try {
            var variant = productService.findVariantById(variantId);
            if (variant != null) {
                String html = """
                    <div class=\"selected-price\">
                        <span class=\"price-label\">Price:</span>
                        <span class=\"price-value\">$%s</span>
                    </div>
                    <div class=\"selected-stock\">
                        <span class=\"stock-label\">Availability:</span>
                        <span class=\"stock-value %s\">%s</span>
                    </div>
                """.formatted(
                        String.format("%.2f", variant.getPrice()),
                        variant.getStockStatus().name().toLowerCase(),
                        switch (variant.getStockStatus().name()) {
                            case "IN_STOCK" -> "In Stock (" + variant.getAvailableQty() + " available)";
                            case "LOW_STOCK" -> "Low Stock (" + variant.getAvailableQty() + " available)";
                            case "OUT_OF_STOCK" -> "Out of Stock";
                            default -> "Unknown";
                        }
                );
                return ResponseEntity.ok(html);
            }
            return ResponseEntity.notFound().build();
        } catch (Exception e) {
            return ResponseEntity.badRequest().build();
        }
    }
}

File: src/main/kotlin/org/example/eshop/controller/ProductController.kt
package org.example.eshop.controller

import org.example.eshop.dto.ProductDto
import org.example.eshop.dto.ProductDetailDto
import org.example.eshop.service.ProductService
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import java.math.BigDecimal

@RestController
@RequestMapping("/api/products")
class ProductController(
    private val productService: ProductService
) {

    @GetMapping
    fun getProducts(
        @RequestParam(required = false) type: String?,
        @RequestParam(required = false) region: String?,
        @RequestParam(required = false) harvestYear: Int?,
        @RequestParam(required = false) minPrice: BigDecimal?,
        @RequestParam(required = false) maxPrice: BigDecimal?,
        @RequestParam(required = false) inStock: Boolean?
    ): ResponseEntity<List<ProductDto>> {
        return try {
            val products = productService.findProducts(
                type = type,
                region = region,
                harvestYear = harvestYear,
                minPrice = minPrice,
                maxPrice = maxPrice,
                inStock = inStock
            )
            ResponseEntity.ok(products)
        } catch (e: IllegalArgumentException) {
            // Handle invalid filter parameters gracefully
            ResponseEntity.badRequest().build()
        }
    }

    @GetMapping("/{slug}")
    fun getProductBySlug(@PathVariable slug: String): ResponseEntity<ProductDetailDto> {
        return productService.findProductBySlug(slug)?.let { product ->
            ResponseEntity.ok(product)
        } ?: ResponseEntity.notFound().build()
    }

    @GetMapping("/variant-info")
    fun getVariantInfo(@RequestParam variantId: Long): ResponseEntity<String> {
        return try {
            val variant = productService.findVariantById(variantId)
            if (variant != null) {
                val html = """
                    <div class="selected-price">
                        <span class="price-label">Price:</span>
                        <span class="price-value">$${String.format("%.2f", variant.price)}</span>
                    </div>
                    <div class="selected-stock">
                        <span class="stock-label">Availability:</span>
                        <span class="stock-value ${variant.stockStatus.name.lowercase()}">
                            ${when (variant.stockStatus.name) {
                                "IN_STOCK" -> "In Stock (${variant.availableQty} available)"
                                "LOW_STOCK" -> "Low Stock (${variant.availableQty} available)"
                                "OUT_OF_STOCK" -> "Out of Stock"
                                else -> "Unknown"
                            }}
                        </span>
                    </div>
                """.trimIndent()
                ResponseEntity.ok(html)
            } else {
                ResponseEntity.notFound().build()
            }
        } catch (e: Exception) {
            ResponseEntity.badRequest().build()
        }
    }
}